### PR TITLE
Perf/finalized txhash mapping encoding

### PR DESCRIPTION
**Pull request by asdacap**
- Add a flag to encode blockNumber instead of blockHash in tx index when block is finalized. Seems to cut down space usage for tx index from 85GB to 50GB.
- Add a flag to not store tx index if block is too old, basically same as geth's `--txlookuplimit`. With the default of about one year, further reduce tx indxe from 50GB to 17.6 GB.

## Changes

- Encode block number instead of blockhash if block is finalized.
- Remove tx index of very old block.
- Moved receipt related config to ReceiptConfig.

#### What types of changes does your code introduce?

- [X] Optimization

#### Requires testing

- [X] Yes
- [ ] No

#### If yes, did you write tests?

- [X] Yes
- [ ] No

#### Notes on testing

- First few million txhash request match archive node. Unable to load enough request to measure consistent difference. Seems to always consume 20ms.

### CODE REVIEW

File: src/Nethermind/Nethermind.Blockchain/Receipts/PersistentReceiptStorage.cs [hunk 1]
```diff
@@ -46,18 +48,63 @@ public class PersistentReceiptStorage : IReceiptStorage
...
+
+                // Delete old tx index
+                if (_receiptConfig.TxLookupLimit > 0)
+                {
+                    Block newOldTx = _blockTree.FindBlock(newMain.Number - _receiptConfig.TxLookupLimit.Value);
+                    if (newOldTx != null)
+                    {
+                        ClearTxIndexForBlock(newOldTx);
+                    }
+                }
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Blockchain/Receipts/PersistentReceiptStorage.cs:87]: Wouldn't it be simpler to just use a block number all the time? And removed reorganized transactions.

Comment (asdacap) [reply src/Nethermind/Nethermind.Blockchain/Receipts/PersistentReceiptStorage.cs:87]: Fair point

File: src/Nethermind/Nethermind.Blockchain/BlockTree.cs [hunk 2]
```diff
@@ -2065,6 +2065,13 @@ internal void ReleaseAcceptingNewBlocks()
             }
         }
 
+        public bool IsFinalized(BlockHeader header)
+        {
+            if (header.Number < _syncConfig.PivotNumberParsed) return true;
+            if (header.Number < Head?.Number - Reorganization.MaxDepth) return true;
```
[the remark this answers is not in the document]
Comment (asdacap) [reply src/Nethermind/Nethermind.Blockchain/BlockTree.cs:2071]: Yea... does not help. CL finalization is a hash, so then to know if a header is finalized (a descendent of it), properly, it'll need to check its child one by one. Might as well, just use reorg number.

[the remark this answers is not in the document]
Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.Blockchain/BlockTree.cs]: we can convert hash to number easily!

**Review by LukaszRozmej — APPROVED** [review]

File: src/Nethermind/Nethermind.Blockchain/BlockTree.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (asdacap) [reply src/Nethermind/Nethermind.Blockchain/BlockTree.cs]: Ah, you mean use the block number of finalized block then.

File: src/Nethermind/Nethermind.Blockchain/Receipts/PersistentReceiptStorage.cs [hunk 3]
```diff
@@ -46,18 +48,58 @@ public class PersistentReceiptStorage : IReceiptStorage
...
         public Keccak FindBlockHash(Keccak txHash)
         {
             var blockHashData = _transactionDb.Get(txHash);
-            return blockHashData is null ? FindReceiptObsolete(txHash)?.BlockHash : new Keccak(blockHashData);
+            if (blockHashData is null) return FindReceiptObsolete(txHash)?.BlockHash;
+
+            if (blockHashData.Length == Keccak.Size) return new Keccak(blockHashData);
+
+            long blockNum = new RlpStream(blockHashData).DecodeLong();
+            return _blockTree.FindBlockHash(blockNum);
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Blockchain/Receipts/PersistentReceiptStorage.cs:102]: One thing here, in a case of the reorg this block might not have this transaction.
We should change this method to FindBlockNumber? And then check it later? But that would be somewhat incompatible with old layout.

Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.Blockchain/Receipts/PersistentReceiptStorage.cs:102]: Ah probably not an issue as we delete reorged tx mappings, still would be good to have a failsafe later if we don't find the transaction in the block?

Comment (asdacap) [reply src/Nethermind/Nethermind.Blockchain/Receipts/PersistentReceiptStorage.cs:102]: ok

File: src/Nethermind/Nethermind.Blockchain/Receipts/PersistentReceiptStorage.cs [hunk 4]
```diff
@@ -46,18 +48,58 @@ public class PersistentReceiptStorage : IReceiptStorage
...
             _migratedBlockNumber = Get(MigrationBlockNumberKey, long.MaxValue);
+
+            _blockTree.BlockAddedToMain += BlockTreeOnBlockAddedToMain;
+        }
+
+        private void BlockTreeOnBlockAddedToMain(object? sender, BlockReplacementEventArgs e)
+        {
+            if (e.PreviousBlock != null)
+            {
+                RemoveBlockTx(e.PreviousBlock);
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Blockchain/Receipts/PersistentReceiptStorage.cs:66]: shouldn't this be under Task.Run?

File: src/Nethermind/Nethermind.Blockchain/Receipts/PersistentReceiptStorage.cs [hunk 5]
```diff
@@ -46,18 +48,58 @@ public class PersistentReceiptStorage : IReceiptStorage
...
-            _blockFinder = blockFinder ?? throw new ArgumentNullException(nameof(blockFinder));
+            _blockTree = blockTree ?? throw new ArgumentNullException(nameof(blockTree));
             _storageDecoder = storageDecoder ?? ReceiptArrayStorageDecoder.Instance;
+            _receiptConfig = receiptConfig ?? throw new ArgumentNullException(nameof(receiptConfig));
 
             byte[] lowestBytes = _database.Get(Keccak.Zero);
             _lowestInsertedReceiptBlock = lowestBytes is null ? (long?)null : new RlpStream(lowestBytes).DecodeLong();
             _migratedBlockNumber = Get(MigrationBlockNumberKey, long.MaxValue);
+
+            _blockTree.BlockAddedToMain += BlockTreeOnBlockAddedToMain;
```
Comment (MarekM25) [inline src/Nethermind/Nethermind.Blockchain/Receipts/PersistentReceiptStorage.cs:59]: Is this event should be turned off if all optimizations are disabled?

File: src/Nethermind/Nethermind.Blockchain/Receipts/PersistentReceiptStorage.cs [hunk 6]
```diff
@@ -222,11 +264,34 @@ public bool HasBlock(Keccak hash)
 
         public void EnsureCanonical(Block block)
         {
-            TxReceipt[] receipts = Get(block);
             using IBatch batch = _transactionDb.StartBatch();
-            foreach (TxReceipt txReceipt in receipts)
+
+            long headNumber = _blockTree.FindBestSuggestedHeader()?.Number ?? 0;
+
+            if (_receiptConfig.TxLookupLimit == -1) return;
```
[the remark this answers is not in the document]
Comment (MarekM25) [reply src/Nethermind/Nethermind.Blockchain/Receipts/PersistentReceiptStorage.cs:271]: ok, it should be fine

**Pull request merged**